bin/create_cluster.sh

```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+if [ $# -eq 1 ];
+then
+  workspace="$1"
+  terraform init
```

Comment:
I don't think this should be interacting with Terraform directly, as AWS infrastructure-level changes should be handled in a separate phase that will run prior to creating or updating clusters. At a high level, we should have the following phases when creating or updating environments:

1. Terraform - update or create main AWS resources such as VPC, subnets, Route53, etc
2. Kops - update or create cluster
3. Helm/kubectl - update or create services on the cluster

The work here should be solely concerned with item no. 2 above, and at a high level do something like this:

1. `terraform output -json` - get values from the current Terraform state that Kops needs to know about
2. `create_cluster_config.py` - pass those values into `create_cluster_config.py` on the command line, (either as individual flags like `--availability-zones`, or the whole tf json object for simplicity's sake), combine it with a template `cluster-template.yml`, and output a final `cluster.yml`
3. `kops create|replace -f cluster.yml` - apply the changes to kops remote config store
4. `kops update` - apply kops remote config to the actual cluster

We don't necessarily need to wrap numbers 1, 3 and 4 above in scripts, unless it's easier to do so, as they could just be steps in our pipeline.

Reply:
Hi, this work in jira was divided into 2 tkts, terraform elements and kops elements
https://dsdmoj.atlassian.net/browse/CPT-472
https://dsdmoj.atlassian.net/browse/CPT-483

regarding the terraform part specifically:

1)Use terraform to provision the VPC and other infrastructure needed.
2)Turn this into an automated job to kick off the terraform build
3)Make it so that any variables required by the kops job in CPT-483 can be passed to it.

that is what create_cluster.sh is doing.
If this is not what we want, please update the described tkts with the correct specification, and ill work on modifying those.

Your description above is just a simplified version of what this pull request is already doing.
If we already have and automated way of creating a cluster, all the involved components, terraform workspace, terraform components (vpc and networking), and kops elements automatically , why turn that into manual steps that require us to handle them?

Reply:
The tickets are correct as far as I can see, and they don't describe handling Terraform as part of cluster creation:

- Use terraform to provision the VPC and other infrastructure needed.
- Turn this into an automated job to kick off the terraform build
- Make it so that any variables required by the kops job in   CPT-483 READY FOR REVIEW  can be passed to it.

Terraform handles management of all AWS resources that _aren't_ the cluster, so there's a separation between what Terraform and Kops do - therefore it makes sense to handle them as separate phases with separate processes and scripts managing them.

Reply:
If that is the case, we can remove that from the create_cluster.sh
This PR also includes changes to the terraform templates to create the VPC and underlying networking parts, from terraform workspaces. So if you create a new terraform workspace and apply the templates it will create the vpc and other parts for the new cluster.
This will be an extra manual step if we remove it from create_cluster.sh

Reply:
this was fixed in https://github.com/ministryofjustice/kubernetes-investigations/pull/22/commits/e81efc201cd5f71f549fffb864a4505d7445b7ab

Reply:
What's the context in which `00-cluster_pipeline.sh` is intended to be executed? I don't think we ever want to execute a single command and create both Terraform and Kops resources all in one go, as they're distinctly different things, and will quite likely have significantly different lifecycles - thoughts @kalbir ?

Reply:
Kerin, 00-cluster_pipeline.sh is just a way of executing the different stepts automatically in a pipeline for 1 cluster.
you can manually execute 01. 02. 03. 04 and accomplish the same

99-multiple_clusters.py is the first script executed by the pipeline, this script determines what clusters need to be created, and then triggers 00-cluster_pipeline.sh for each cluster, witch internally executes the steps for each cluster, 
01-terraform_resources.sh
02-rsa_key.sh
03-cluster_config.py
04-kops_elements.sh

Reply:
OK - but that means that the Terraform phase and cluster phase are still bound together, and any change will trigger both `terraform apply` and `kops create`. We don't want that, as they have separate lifecycles - we're likely to have frequent changes to Terraform resources that don't impact the cluster at all, so they shouldn't trigger anything with Kops.

Similarly, the Kops `03-cluster_config.py` script is still interacting with Terraform directly, which binds the two together. Kops does not need to know anything about Terraform, so it would be cleaner to pass information obtained from Terraform into this script as command line arguments, as per [previous feedback](03-cluster_config.py) - this would make the script more flexible, as it wouldn't depend on any extra tooling or particular execution context.